src/sentry/ingest/ingest_consumer.py

```diff
@@ -105,6 +67,83 @@ def shutdown(self):
         pass
 
 
+def process_event(message):
+    payload = message["payload"]
+    start_time = float(message["start_time"])
+    event_id = message["event_id"]
+    project_id = message["project_id"]
+    remote_addr = message.get("remote_addr")
+    attachments = message.get("attachments") or ()
+
+    # check that we haven't already processed this event (a previous instance of the forwarder
+    # died before it could commit the event queue offset)
+    deduplication_key = "ev:{}:{}".format(project_id, event_id)
+    if cache.get(deduplication_key) is not None:
+        logger.warning(
+            "pre-process-forwarder detected a duplicated event" " with id:%s for project:%s.",
+            event_id,
+            project_id,
+        )
+        return  # message already processed do not reprocess
+
+    try:
+        project = Project.objects.get_from_cache(id=project_id)
+    except Project.DoesNotExist:
+        logger.error("Project for ingested event does not exist: %s", project_id)
+        return
+
+    # Parse the JSON payload. This is required to compute the cache key and
+    # call process_event. The payload will be put into Kafka raw, to avoid
+    # serializing it again.
+    # XXX: Do not use CanonicalKeyDict here. This may break preprocess_event
+    # which assumes that data passed in is a raw dictionary.
+    data = json.loads(payload)
+
+    cache_key = cache_key_for_event(data)
+    default_cache.set(cache_key, data, CACHE_TIMEOUT)
+
+    if attachments:
+        attachment_cache.set(
+            cache_key,
+            attachments=[
+                CachedAttachment(
+                    type=attachment.pop("attachment_type"), meta_only=True, **attachment
+                )
+                for attachment in attachments
+            ],
```

Comment:
_nit_: Please move this list into a variable to get nicer formatting.

Reply:
done